### ISSUE TITLE
Feign is creating a relative url with http://

### ISSUE THREAD

**Issue by leonp967** [OpenFeign/feign#996]
Hello, I'm implementing 3 micro services with Spring Boot, and one of them calls the other 2 to reply with the desired response. In this one, I need to use Feign together with Eureka and Ribbon, but without Spring Cloud. For that, I configured the registration of the micro services in Eureka manually, and they are registering successfully. In the third service, I'm querying these services URLs to be able to call them, and with the URL returned by the Eureka client, I create a Feign client to call the service, but I'm having a little problem with the way Feign is resolving the request, which is giving me an exception, I'll explain what exactly I'm doing.

Here is where I'm creating the Feign client:
```java
    @Bean
    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
    public PlaylistServiceClient playlistServiceClient(String serviceUrl){
        return Feign.builder()
                .client(RibbonClient.create())
                .encoder(new JacksonEncoder())
                .decoder(new JacksonDecoder())
                .logger(new Slf4jLogger(PlaylistServiceClient.class))
                .logLevel(Logger.Level.FULL)
                .target(PlaylistServiceClient.class, serviceUrl);
    }
```
When I call that method, I'm passing the URL returned from Eureka, which looks like this: `http://localhost:9003`, and my client definition is this:
```java
public interface PlaylistServiceClient {

    @RequestLine("GET /playlist-service/playlist/{id}")
    PlaylistResponse getPlaylist(@Param("id") String id);
}
```
When I call the `getBean()` method from Spring, the client is created successfully, but the exception is thrown when I call the `getPlaylist()` method from the just created client. I debugged the execution on the Feign code, and I found where the exception is thrown, it's in this piece of code:
```java
public RequestTemplate uri(String uri, boolean append) {
    /* validate and ensure that the url is always a relative one */
    if (UriUtils.isAbsolute(uri)) {
      throw new IllegalArgumentException("url values must be not be absolute.");
    }
```
This method is located on the RequestTemplate class from the feign-core library, and from what I could see, the problem is that in this point of execution, the `uri` parameter has this value: `http://:9003/playlist-service/playlist/bc40c33b-d1b2-4e0c-9ad8-8ec424abbe89`. I think that the "localhost" part was removed from the URL to create a relative URL, since both services are running on my machine, but the `UriUtils.isAbsolute()` method recognizes any URL that is not null and not empy, that starts with "http://", as an absolute URL, and then the exception is thrown. I could not find exactly which method removed the "localhost" from the original URL, but I think it should have removed the "http://" part too. 

Is this behavior expected? If so, how can I workaround this?

**Comment by glani** [comment]
Have you tried to do the same via ip ? I had the same issue with spring cloud however the problem that resolver got the wrong url.

**Comment by leonp967** [comment]
@glani Yes, I've tried with the IP and the hostname, and the same exception is thrown

**Comment by leonp967** [comment]
For anyone who stumbled upon the same problem, I managed to solve it. It turns out I was using Feign + Ribbon + Eureka wrongly (no surprises in this part). The correct way to make it work, or at least the way I was able to call my other services, is to set some properties on the ``application.properties`` of my service that calls the others, and load the ``application.properties`` resource file with the ``ConfigurationManager.loadPropertiesFromResources()``. That way the Ribbon client will use the properties defined in the file and correctly discover your services by their VipAddress registered on Eureka.

The properties I had to set were these:
```
eureka.serviceUrl.default=your eureka url
eureka.client.fetchRegistry=true
ribbon.eureka.enabled=true
"yourClientName".ribbon.DeploymentContextBasedVipAddresses=the service's vipAddress
ribbon.NIWSServerListClassName=com.netflix.niws.loadbalancer.DiscoveryEnabledNIWSServerList
```
And when I create the Feign client, I pass as parameter the same client name used in the .properties file. Now it's working greatly.
